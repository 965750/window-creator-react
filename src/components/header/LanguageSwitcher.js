import React, { Component } from 'react'
import { connect } from 'react-redux'
import { setLocalLang } from '../../store/actions/themeActions'
import PropTypes from 'prop-types'
import styles from './LanguageSwitcher.module.scss'
import { FormattedMessage, injectIntl } from 'react-intl'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faChevronDown } from '@fortawesome/free-solid-svg-icons'

class LanguageSwitcher extends Component {
    state = {
        languages: [
            {
                value: 'pl',
                label: 'Polish',
            },
            {
                value: 'en',
                label: 'English',
            },
        ],
    }
    changeSelect = e => {
        this.props.setLocalLang(e.target.value)
    }

    render() {
        return (
            <div className="flex absolute top-25 right-10">
                <p className="leading-loose">
                    <FormattedMessage
                        id="Select language:"
                        defaultMessage="Select language:"
                    />
                </p>
                <div
                    className={`${styles.select__wrapper} relative relative ml-5`}
                >
                    <FontAwesomeIcon
                        className={`${styles.select__arrow} absolute`}
                        icon={faChevronDown}
                    />
                    <select
                        onChange={this.changeSelect}
                        value={this.props.language}
                        className={`${styles.select} h-full w-full pl-3 border border-gullGray rounded bg-white outline-none`}
                    >
                        {this.state.languages.map(option => {
                            return (
                                <option value={option.value} key={option.value}>
                                    {this.props.intl.formatMessage({
                                        id: option.label,
                                    })}
                                </option>
                            )
                        })}
                    </select>
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => {
    return {
        language: state.theme.lang,
    }
}

const mapDispatchToProps = dispatch => {
    return {
        setLocalLang: lang => dispatch(setLocalLang(lang)),
    }
}

LanguageSwitcher.propTypes = {
    setLocalLang: PropTypes.func,
}

export default injectIntl(
    connect(
        mapStateToProps,
        mapDispatchToProps
    )(LanguageSwitcher)
)
