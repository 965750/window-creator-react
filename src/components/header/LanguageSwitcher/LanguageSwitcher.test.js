import '@testing-library/jest-dom'

import { applyMiddleware, compose, createStore } from 'redux';
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { render, cleanup, fireEvent } from "@testing-library/react"

import LanguageSwitcher from './index'
import messages from "../../../messages-i18n"
import thunk from 'redux-thunk';
import React from 'react'

const initialStore = {
  theme: {
    lang: 'pl'
  }
}
const reducer = (state = initialStore, action) => {
  switch (action.type) {
    case 'SET_LOCAL_LANG':
      return {
        ...state,
        theme: {
          ...state.theme,
          lang: 'en'
        },
      }
    default:
      return state
  }
}

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
  ),
)

const renderWithRedux = (component) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>)
  }
}

describe('LanguageSwitcher', () => {
  let wrapper

  beforeEach(() => {
    wrapper = renderWithRedux(
      <IntlProvider
        locale='en'
        messages={messages[store.getState().theme.lang]}
      >
        <LanguageSwitcher />
      </IntlProvider>
    )
  })

  afterEach(cleanup)

  it('should render without errors', () => {
    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should be set to pl as initial language', () => {
    wrapper.getByText(/Wybierz jÄ™zyk/)
  })

  it('should change selected language', () => {
    const select = wrapper.getByTestId('select')

    expect(wrapper.getByText('Polski')).toBeTruthy();

    fireEvent.change(select,{ target: {value: 'en'}})

    cleanup()

    wrapper = renderWithRedux(
      <IntlProvider
        locale='en'
        messages={messages[store.getState().theme.lang]}
      >
        <LanguageSwitcher />
      </IntlProvider>
    )

    expect(wrapper.getByText('Polish')).toBeTruthy();
  })
})
