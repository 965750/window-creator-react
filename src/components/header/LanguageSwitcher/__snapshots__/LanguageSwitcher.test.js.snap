// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSwitcher should render without errors 1`] = `
<DocumentFragment>
  <div
    class="flex"
  >
    <p
      class="select__label mr-5"
    >
      Wybierz jÄ™zyk:
    </p>
    <div
      class="select__wrapper relative relative mt-lg"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-down fa-w-14 select__arrow absolute right-10"
        data-icon="chevron-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
          fill="currentColor"
        />
      </svg>
      <select
        class="select h-full w-full pl-3 border border-gullGray rounded bg-white outline-none"
        data-testid="select"
      >
        <option
          value="pl"
        >
          Polski
        </option>
        <option
          value="en"
        >
          Angielski
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
